shell: Redesign untrusted "add host" dialog

martinpitt opened the pull request:
Remote machines can send malicious cockpit pages which do iframe traversal or use the `host` channel option to run arbitrary code on all other currently connected Cockpit sessions.

Isolating cockpit sessions from different hosts from one another is difficult, and properly implementing this will take a while (if it's possible at all): https://issues.redhat.com/browse/COCKPIT-870

Until this happens, warn users about that in the "unknown host" shell dialog. 

Also hide the "verify fingerprint" recipe behind an expander. In practice, most users do "Trust On First Use" anyway, as validating a FP properly is too hard. This makes the dialog less overwhelming and also puts more emphasis on the "trust" aspect.

Add a check to TestMultiMachine.testTroubleshooting which actually runs the generated validation command and ensures that the shown fingerprint matches.

----

Initial dialog, FP validation collapsed by default:

![image](https://github.com/cockpit-project/cockpit/assets/200109/9a152020-989f-48bc-884a-ff4c8a0a1561)

FP validation expanded:

![image](https://github.com/cockpit-project/cockpit/assets/200109/8bc9a486-0da0-4706-ad4a-963d49a2f9b6)

FP popover:

![image](https://github.com/cockpit-project/cockpit/assets/200109/0acc4ca3-c64b-42a9-9709-65fdfbb83193)

martinpitt commented:
If possible, I'd also like to trim the alert text a bit, to fit on one line (at least in English). Perhaps:

> Malicious remote Web Console pages can (influence|change|damage) other connected hosts in this session.

?

garrett commented:
This modal is way overloaded. It's not clear what's going on, what's important, what someone needs to do, etc. I'm trying to figure this out in a mockup right now.

As for the new part of this PR, the security issue: It's not error, nor is it an alert (of something that's been done already). Opting in is clicking the button; it doesn't need to have an "unbreak the UI" toggle.

Still, this modal was already too complex before adding anything else, and it's extremely overloaded now. I hope to finish up a suggestion with a mockup soon and I'll share it here when I have it.

garrett commented:
@martinpitt: Ah, that's what I thought, but wasn't sure. Additionally, corps should really be using CA with SSH certs, but in practice, few do.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/sec-creating_ssh_ca_certificate_signing-keys

https://smallstep.com/blog/use-ssh-certificates/

I guess I'll redesign the modal to be even simpler now!

garrett commented:
This isn't done, but I thuoght I should share what I have before the upcoming long weekend. I'd like to get feedback.

HOSTNAME is the actual hostname or IP address.

![image](https://github.com/cockpit-project/cockpit/assets/10246/442e5597-930c-49f6-9455-b1c7bfd33de7)

I don't have the "View details" popover for the details for the session stuff yet. We might want to have a popover for "certificate authority" and we might want to ditch the extended sharing info.

What do you think so far?

garrett commented:
Oh, the warning text at the bottom has a typo. I revised it a few times and accidentally left it in that state. :sweat_smile:

martinpitt commented:
Nice, thanks @garrett ! I like the decomposition/expanders. I'm less sure about the "not using certificate authority"  -- that's one way to assert a known host, but (unfortunately) more of a fringe case, and confusing for casual users. I like the  original "you are connecting to hostname for the first time", as when you see it, it's exactly right, and doesn't induce fear.

martinpitt commented:
@garrett First iteration towards your design. Initial dialog (not expanded):

![image](https://github.com/cockpit-project/cockpit/assets/200109/be723de4-a25e-4298-ba20-61219decaa4a)

Expanding:

![image](https://github.com/cockpit-project/cockpit/assets/200109/0310a4a3-23e5-4871-a5c9-9eb53b6a2e96)

Opening popup:

![image](https://github.com/cockpit-project/cockpit/assets/200109/dfadedca-743c-4934-9781-14d64b3920b3)

Diff between design and implementation:
 - Does not speak about "certificate authority", see above. I kept the original "for the first time" text, as that's the 95% case.
 - Drop the redundant "Verify the fingerprint by [running command..]", as the heading already says that.

allisonkarlitskaya commented on pkg/shell/hosts_dialog.jsx:
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
+                    <p>{_("The fingerprint should match:")} {fingerprint_help}</p>
+                    <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help pf-v5-u-font-family-monospace">{fp}</ClipboardCopy>
+                </ExpandableSection>
+                <Alert variant='warning' isInline isPlain title={_("Malicious pages on the remote machine can do arbitrary changes to other connected hosts in the same session.")} />
```
`s/can do/can make/`

martinpitt replied:
Superseded by Garrett's proposal below.

garrett commented:
These changes work for me! 

This seems very verbose and hard to follow:

> Malicious pages on the remote machine can do arbitrary changes to other connected hosts in the same session.

What I shared:

> Pages from the remote machine can _make changes_ to other hosts in the same session. 

(Emphasis for the part I accidentally left out in the mockup. Whoops!)

What about this?

_Malicious pages on a remote system may affect other connected hosts._

martinpitt commented:
Thanks @garrett and @allisonkarlitskaya ! I updated the warning string, and put current screenshots into the description. Triggering f38 tests to see what I need to adjust.

cockpituous commented on pkg/shell/hosts_dialog.jsx:
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
                 <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help-cmds pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
                 <p>{_("The resulting fingerprint is fine to share via public methods, including email.")}</p>
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <p>{_("If the fingerprint matches, click 'Trust and add host'. Otherwise, do not connect and contact your administrator.")}</p>
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231010-153219-d8bb31b4-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt replied:
I added test for this now.

cockpituous commented on pkg/shell/hosts_dialog.jsx:
```diff
@@ -479,18 +484,27 @@ class HostKey extends React.Component {
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <ExpandableSection toggleText={ _("Verify fingerprint") }
+                                   isExpanded={this.state.verifyExpanded}
+                                   onToggle={(_ev, verifyExpanded) => this.setState({ verifyExpanded }) }>
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231010-153219-d8bb31b4-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt replied:
Added a test.

martinpitt commented:
This is now ready. I did the final tweaks discussed with @garrett (drop the period in the alert), adjusted tests, and cleaned them up a bit while I was at it.

cockpituous commented on pkg/shell/hosts_dialog.jsx:
```diff
@@ -479,18 +484,24 @@ class HostKey extends React.Component {
                 <ClipboardCopy isReadOnly hoverTip={_("Copy")} clickTip={_("Copied")} className="hostkey-verify-help-cmds pf-v5-u-font-family-monospace">ssh-keyscan -t {key_type} localhost | ssh-keygen -lf -</ClipboardCopy>
                 <p>{_("The resulting fingerprint is fine to share via public methods, including email.")}</p>
-                <p>{_("If the fingerprint matches, click 'Accept key and connect'. Otherwise, do not connect and contact your administrator.")}</p>
+                <p>{_("If the fingerprint matches, click 'Trust and add host'. Otherwise, do not connect and contact your administrator.")}</p>
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19409-20231011-054038-2287fc87-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

jelly reviewed (approved):
Nice improvements!

garrett reviewed (approved):
Thanks! It's a lot better now! :+1: Ready to merge. :tada:
